### PR TITLE
[master] detach() fixes

### LINKED ISSUE #2017: Lazy collections are fetched on detach
I've recently migrated from javax to jakarta, from Eclipselink 2.x to 4.0.2.
Something that didn't happen before, but is happening with 4.0.2 is that when calling `em.detach(entity)`, if the entity has an uninitialized lazy collection, it gets fetched in the detach operation.
This is a high impact bug, as it could potentially load fetch a lot of db rows to memory.
In the project I work, we have many mapped `@OneToMany` collections which are private, and are only there to cascade deletes.
And we do a lot of detach to prevent many objects from end up in memory.
With this bug, the result is the exact opposite...

I have created a minimal project to reproduce the issue: https://github.com/luisfpg/eclipselink-detach-bug

I'm copying here parts of the README to make it easier to screen the problem:

This is the code to reproduce the issue:

```
var state = em.find(State.class, "s1");
var cities = (IndirectSet<City>) state.getCities();
System.out.printf("Is collection initialized? %s%n", cities.isInstantiated());
System.out.println("Before detach");
em.detach(state);
System.out.printf("State %s detached%n", state.getCode());
System.out.printf("Is collection initialized? %s%n", cities.isInstantiated());
```

And the output is:

```
[EL Info]: 2023-12-06 08:02:59.36--ServerSession(232307208)--EclipseLink, version: Eclipse Persistence Services - 4.0.2.v202306161219
Loaded state s1
Is collection initialized? false
Before detach
Loaded state s1
Loaded city s1b
Loaded city s1a
State s1 detached
Is collection initialized? true
```

Note that the collection gets initialized by the detach operation, which makes no sense.

### PR DESCRIPTION
**Pull request by rfelcman**
There are two fixes for EntityManager.detach():
- Lazy collections are fetched on detach
  Unnecessary fetch of lazy loaded collection `Indirect...` declared in the detached entity. See #2017
- StackOverflowError because of cyclic DETACH
  It happens if there is cross relation between entities like 
  `@OneToMany(cascade = {CascadeType.ALL }, fetch = FetchType.LAZY,...` from the first entity and
  `@ManyToOne(cascade = { CascadeType.ALL}, fetch = FetchType.EAGER,...` from the second. See #1577
fixes #2017 #1577 
replaces #292

### CODE REVIEW

File: foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/descriptors/DescriptorIterator.java [hunk 1]
```diff
@@ -92,6 +94,7 @@ public abstract class DescriptorIterator {
     protected DescriptorIterator() {
         // 2612538 - the default size of Map (32) is appropriate
         this.visitedObjects = new IdentityHashMap();
+        this.visitedMappings = new HashMap();
```
Comment (lukasj) [inline foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/descriptors/DescriptorIterator.java:97]: use generics

File: foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/descriptors/DescriptorIterator.java [hunk 2]
```diff
@@ -441,7 +455,9 @@ protected void internalIterateReferenceObjects(Object sourceObject) {
             this.currentItem = currentItemOriginal;
         } else {
             for (DatabaseMapping mapping : mappings) {
-                mapping.iterate(this);
+                if (!this.cascadeCondition.shouldNotCascade(mapping)) {
```
Comment (lukasj) [inline foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/descriptors/DescriptorIterator.java:458]: I'd probably rather add `shouldCascade` to the cascadeCondition

File: foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/descriptors/DescriptorIterator.java [hunk 3]
```diff
@@ -707,6 +707,10 @@ public class CascadeCondition{
         public CascadeCondition() {
         }
 
+        public boolean shouldCascade(DatabaseMapping mapping){
+            return !shouldNotCascade(mapping);
```
Comment (lukasj) [inline foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/descriptors/DescriptorIterator.java:711]: the whole point of my comment was to avoid negation of negated value..., the idea is for this to be:
```
return shouldCascadeAllParts() || (shouldCascadePrivateParts() && mapping.isPrivateOwned())
```

and `shouldNotCascade` to be:
```
return !shouldCascade(...);
```

so in the end one can avoid having a call stack evaluated to:
`(!(!(shouldCascadeAllParts() || (shouldCascadePrivateParts() && mapping.isPrivateOwned())))`

Comment (rfelcman) [reply foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/descriptors/DescriptorIterator.java]: Uff fixed. Needed update `UnitOfWorkImpl` too.

**Review by lukasj — APPROVED** [review]

**Comment by chicko12345** [comment]
Hello Community,

as i understand this fix wasn't back ported for 4.x. Am i right?

Kind Regards

**Comment by rfelcman** [comment]
> Hello Community,
> 
> as i understand this fix wasn't back ported for 4.x. Am i right?
> 
> Kind Regards

See
https://github.com/eclipse-ee4j/eclipselink/pull/2225

**Comment by chicko12345** [comment]
> > Hello Community,
> > as i understand this fix wasn't back ported for 4.x. Am i right?
> > Kind Regards
> 
> See #2225

Thank you very much!

**Comment by rfelcman** [comment]
> > > Hello Community,
> > > as i understand this fix wasn't back ported for 4.x. Am i right?
> > > Kind Regards
> > 
> > 
> > See #2225
> 
> Thank you very much!

Hello,
it's available in the latest snapshots e.g. https://jakarta.oss.sonatype.org/content/groups/staging/org/eclipse/persistence/eclipselink/4.0.5-SNAPSHOT/ but it's not included in the latest release 4.0.4 https://github.com/eclipse-ee4j/eclipselink/releases/tag/4.0.4 . Will be available in 4.0.5 release (autumn 2024).

**Comment by chicko12345** [comment]
> > > > Hello Community,
> > > > as i understand this fix wasn't back ported for 4.x. Am i right?
> > > > Kind Regards
> > > 
> > > 
> > > See #2225
> > 
> > 
> > Thank you very much!
> 
> Hello, it's available in the latest snapshots e.g. https://jakarta.oss.sonatype.org/content/groups/staging/org/eclipse/persistence/eclipselink/4.0.5-SNAPSHOT/ but it's not included in the latest release 4.0.4 https://github.com/eclipse-ee4j/eclipselink/releases/tag/4.0.4 . Will be available in 4.0.5 release (autumn 2024).

Tnx for the info. This one was a show stopper. So if it'll be available soon (autumn is absolutely ok) - very nice.

**Comment by chicko12345** [comment]
Hi Community,

since the fix is now an official part of glassfish 7.0.22 we've decided to upgrade. For the small datasets it did the trick. However, during testing with bigger datasets (realistic cases in production environments) we run into the same stack overflow issue. Should i create a new issue?

**Comment by rfelcman** [comment]
> Hi Community,
> 
> since the fix is now an official part of glassfish 7.0.22 we've decided to upgrade. For the small datasets it did the trick. However, during testing with bigger datasets (realistic cases in production environments) we run into the same stack overflow issue. Should i create a new issue?

Yes please, but we need what is mentioned behind "bigger datasets". The best could be some test case which leads into issue.
Or is it possible to simulate it by https://github.com/eclipse-ee4j/eclipselink/pull/2037/files#diff-016938172968e33273de6874cb7b991a500d4682b7bdf1c226651ab92b57a8a1
but with more DB rows/entities?
Or there is different entity model/tree?
